pkg/apis/workflow/v1alpha1/workflow_phase.go

```diff
@@ -10,6 +10,7 @@ const (
 	WorkflowSucceeded WorkflowPhase = "Succeeded"
 	WorkflowFailed    WorkflowPhase = "Failed" // it maybe that the workflow was terminated
 	WorkflowError     WorkflowPhase = "Error"
+	WorkflowCanceled  WorkflowPhase = "Canceled" // it is an intermediate state when enable failFast. Workflow phase will be changed from Canceled to Succeeded/Failed/Error
```

Comment:
I am not sure if we need a new phase for this since it's a one-off use case

Reply:
> I am not sure if we need a new phase for this since it's a one-off use case

This is indeed a controversial place, this status just is used to record that this workflow has failed fast. If this status is unreasonable, I can also add `hasFailedFast` field in `WorkflowStatus` struct. Do you have any suggestions

Reply:
@terrytangyuan I'd love for this to become a permanent phase as requested in #11849.  This change would allow for that, as a side effect.  If we could leave as an actual phase, in another PR we could update the UI and API to allow filtering of this phase. 🤔

Reply:
Introducing a new phase would be a much bigger change with potential side effects since we would need to go through the codebase to make sure we handle the phases correctly.

Reply:
Understood. Just a brief look on my own has shown it is a substantial change. Still might take a stab at it. 😃 will see.

Reply:
From my personal understanding of the source code logic, adding such a phase can be seen from two perspectives. If it is only an intermediate state, it will not actually have an impact. If it is the final state, it requires comprehensive code testing.

So what should we do next about this pr? How about add `hasFailedFast` field in `WorkflowStatus` struct? I need a field in status to indicate that the current workflow needs to fail fast

Reply:
Yea I did mention it would require a sizeable refactor in https://github.com/argoproj/argo-workflows/issues/11849#issuecomment-1728009697. It should almost certainly be its own change.

RE: `failFast`, I don't think it would be an appropriate place for a _Workflow-level_ `Cancelled` status either. `failFast` could potentially use a _Node-level_ `Cancelled` status (as opposed to `Skipped`) as mentioned in https://github.com/argoproj/argo-workflows/issues/10312#issuecomment-1760478799.



> So what should we do next about this pr? How about add `hasFailedFast` field in `WorkflowStatus` struct? I need a field in status to indicate that the current workflow needs to fail fast

Are you unable to infer the status without such a field? If `failFast` is `true` or unset, and a node in a branch of the DAG has failed, then we can infer that the rest of the Workflow should fail fast.

Reply:
> Are you unable to infer the status without such a field? If `failFast` is `true` or unset, and a node in a branch of the DAG has failed, then we can infer that the rest of the Workflow should fail fast.

At this time the logic of DAG is a loop: `ExecuteDAG` ->`AssessDAGPhase` ->End/Continue running. A BFS traversal and the result calculation of the target tasks has already been done in `AssessDAGPhase`. My design is to save the results of the last traversal to a field in the status, so that the next loop will directly obtain the fastFail judgment result, without the need for another iteration and calculation of the target tasks.

I want to know which design you prefer so that I can make further modifications

Reply:
@agilgur5 I have modified logic, please review the logic code again